kano_updater/monitor.py

```diff
@@ -0,0 +1,165 @@
+# monitor.py
+#
+# Copyright (C) 2014-2018 Kano Computing Ltd.
+# License: http://www.gnu.org/licenses/gpl-2.0.txt GNU GPL v2
+#
+# Module to allow us to detect when a process is stuck.
+
+# Also heartbeat() function for a monitored process
+# to notify the monitor that it is proceeding normally.
+
+import subprocess
+from collections import defaultdict
+import sys
+import time
+import signal
+import os
+from kano.logging import logger
+import kano_updater.utils
+
+MONITOR_TIMEOUT = 20 * 60
+TIMEOUT_RC = 105
+
+
+class monitorPids:
+    """
+    Class for monitoring a subprocess tree. If the (recursive) set of child
+    processes changes, we assume it is still making progress.
+    """
+    def __init__(self, top_pid):
+        self.top_pid = top_pid
+        self.curr_children = set()
+
+    def _get_children(self):
+        """
+        return a set() of all children (recursively) of
+        self.top_pid (inclusive)
+        """
+        try:
+            stdout = subprocess.check_output(["/bin/ps", "-eo", "ppid,pid"])
+            coll = defaultdict(set)
+
+            for line in stdout.split('\n'):
+                items = line.split()
+                if len(items) >= 2:
+                    ppid, pid = items[:2]
+                    if ppid == 'PPID':
+                        continue
+            try:
+                coll[int(ppid)].add(int(pid))
+            except:
+                pass  # ignore int conversion failure
+
+            def closure(pids):
```

Comment:
While this may be a closure, that name isn't particularly helpful to understand what is going on - maybe call it something like `get_child_pids()` or something (although that might conflict with the main function name. Also a docstring would be helpful to understand the arguments required.

Reply:
Actually this function is performing a transitive closure. It's not a closure itself except in a trivial sense. But I'll rename it to transitive_closure.